### ISSUE TITLE
Intermittent booting into bootloader mode

### ISSUE THREAD

**Issue by charlespax** [PaxInstruments/ATmega32U4-bootloader#11]
A user reported : "Occasionally, my T400 powers up almost immediately, but most of the time the display stays blank for about 20 seconds after pressing the power button, before the backlight comes on and the unit starts working. It this the way it's supposed to work?" See http://community.paxinstruments.com/t/t400-boot-up-time/159

Bootloader mode is normally entered by holding down buttons D and E bringing logic low. Internal pullup resistors are used to set the default state high.
- It could be that the pullup resistors are not enabled before checking the values of D and E. D and E would be floating.
- The bootloader checks the value of D and E too early. If the check happens before the length of the rise time it could read logic low

Here is a code block from `main` in https://github.com/PaxInstruments/ATmega32U4-bootloader/blob/master/avr/bootloaders/caterina/Caterina.c.

```
#if DEVICE_PID == 0x1000 // Only on the T400
        // Set the PWR_ONOFF pin to a low output now, so that it is at a known value.
        // PWR_ONOFF pin should be on PD5 in Rev 13
        PORTD   &= ~(_BV(PORTD5));
        DDRD    |= _BV(DDB5);

    // Put the user buttons D (PB5) and E (PB6) in input mode, so we can read the button state
    DDRB    &= ~(_BV(DDB5) | _BV(DDB6));
#endif

    if (mcusr_state & (1<<EXTRF)) {
        // External reset -  we should continue to self-programming mode.
#if DEVICE_PID == 0x1000   // Only on the T400
    } else if ((PINB & (_BV(PINB5) | _BV(PINB6))) == 0) {
        // User buttons D (PB5) and E (PB6) held low- stay in bootloader mode
#endif
    } else if ((mcusr_state & (1<<PORF)) && (pgm_read_word(0) != 0xFFFF)) {     
        // After a power-on reset skip the bootloader and jump straight to sketch 
        // if one exists.   
        StartSketch();
    } else if ((mcusr_state & (1<<WDRF)) && (bootKeyPtrVal != bootKey) && (pgm_read_word(0) != 0xFFFF)) {   
        // If it looks like an "accidental" watchdog reset then start the sketch.
        StartSketch();
    }
```

It looks like D and E are set as inputs then immediately checked. I don't see the internal pullups enabled.

If this is what is causing the problem, we have two possible fixes: change the bootloader or add external pullup resistors. On the manufacturing side the bootloader should be updated. On the user side users can choose to update the bootloader or to install the pullup resistors.

**Comment by charlespax** [comment]
To turn on internal pullup resistors on PB5 and PB6 use `PORTB = 0b01100000` after setting PORTB to input mode.

**Comment by cibomahto** [comment]
Stylistically, this is a little more specific:
PORTB |=  _BV(PORTB5) | _BV(PORTB6);

It's preferred to use the |= or &= operators when modifying port values like this, to prevent accidentally modifying other bits that aren't relevant to the current operation.
The _BV(n) macro expands to (1<<n), and the PORTBx values should be defined to their specific bit (PORTB0 = 0, etc), so it ends up equaling that binary value. It's usually a little clearer to read when reviewing it later, since it uses symbols to describet intention.

**Comment by charlespax** [comment]
Fixed in commit c4c5f92398ab61197d6aec0f4bd6f3151bbe89e0
